presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionarySelectiveReader.java

```diff
@@ -365,33 +386,69 @@ public Block getBlock(int[] positions, int positionCount)
         }
 
-        if (positionCount == outputPositionCount) {
-            DictionaryBlock block = new DictionaryBlock(positionCount, dictionary, values);
-            values = null;
-            return block;
+        // compact values(ids) array, and calculate 1) the slice sizeInBytes if materialized, and 2) number of nulls
+        long blockSizeInBytes = 0;
+        int nullsCount = 0;  // the nulls count for selected positions
+        int i = 0;
+        int j = 0;
+        while (i < positionCount && j < outputPositionCount) {
+            if (positions[i] != outputPositions[j]) {
+                j++;
+                continue;
+            }
+
+            int id = this.values[j];
+            values[i] = id;
+
+            blockSizeInBytes += dictionaryOffsetVector[id + 1] - dictionaryOffsetVector[id];
+            nullsCount += id == currentDictionarySize - 1 ? 1 : 0;
+
+            i++;
+            j++;
         }
 
-        int[] valuesCopy = new int[positionCount];
+        // If all selected positions are null, just return RLE block.
+        if (nullsCount == outputPositionCount) {
+            return new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), positionCount);
+        }
 
-        int positionIndex = 0;
-        int nextPosition = positions[positionIndex];
-        for (int i = 0; i < outputPositionCount; i++) {
-            if (outputPositions[i] < nextPosition) {
-                continue;
+        // If the expected materialized size of the output block is smaller than a certain ratio of the dictionary size, we will materialize the values
+        int dictionarySizeInBytes = dictionaryOffsetVector[currentDictionarySize - 1];
+        if (blockSizeInBytes * BATCHES_PER_ROWGROUP < dictionarySizeInBytes / MATERIALIZATION_RATIO) {
+            byte[] sliceData = new byte[toIntExact(blockSizeInBytes)];
+            int[] offsetVector = new int[positionCount + 1];
+            int currentOffset = 0;
+            for (int k = 0; k < positionCount; k++) {
+                int id = values[k];
+                int offset = dictionaryOffsetVector[id];
+                int length = dictionaryOffsetVector[id + 1] - offset;
+                System.arraycopy(dictionaryData, offset, sliceData, currentOffset, length);
+
+                currentOffset += length;
+                offsetVector[k + 1] = currentOffset;
             }
-            assert outputPositions[i] == nextPosition;
-            valuesCopy[positionIndex] = this.values[i];
-            positionIndex++;
-            if (positionIndex >= positionCount) {
-                break;
+
+            if (nullsCount > 0) {
+                boolean[] isNullVector = new boolean[positionCount];
+                for (int k = 0; k < positionCount; k++) {
+                    if (values[k] == currentDictionarySize - 1) {
+                        isNullVector[k] = true;
+                    }
+                }
```

Comment:
do you think we could extract some of this into functions and have some unit tests for them? there's a lot going on and it is easy to make mistake. Extracting into a function whose name describes what we're doing would also make it more readable :) What do you think? (L394- L411, and L422-L32.. etc)

Reply:
@sujay-jain Thanks for the suggestion. I extracted getMaterializedBlock() since it's a relatively self-explanatory unit. L394- L411 was just a rewrite of original L380-L393 which was inlined in getBlock(). I think it's easier for readers to understand if we keep it that way. This piece is not as isolated as getMaterializedBlock(), and jumping between the methods would make readers have context switches back and forth. 

Our testing framework uses OrcTester and tests round trips and match the input and output. The new code logic is already covered in `TestSelectiveOrcReader`. I updated the comments in `testVarchars()` there. There is no testing framework that can test each individual functions and building one from ground is beyond this PR. We can't just call getBlock() by itself without first calling read(). And if we call both, it's the same as what the current OrcTester does.